File: src/algorithms/arrays/cyclic-sort/cyclic-sort/sources/cyclic-sort.py
# Cyclic Sort — O(n) sort for arrays containing values 1..n by placing each at index value-1
def cyclic_sort(input_array: list[int]) -> list[int]:
    result = input_array[:]  # @step:initialize
    current_index = 0  # @step:initialize

    while current_index < len(result):
        current_value = result[current_index]  # @step:compare
        correct_index = current_value - 1  # @step:compare

        if current_value != current_index + 1:  # @step:compare
            result[correct_index], result[current_index] = result[current_index], result[correct_index]  # @step:swap
        else:
            current_index += 1  # @step:visit

    return result  # @step:complete

File: src/algorithms/arrays/cyclic-sort/cyclic-sort/sources/cyclic-sort.go
// Cyclic Sort — O(n) sort for arrays containing values 1..n by placing each at index value-1
package cyclicsort

func cyclicSort(inputArray []int) []int {
	result := make([]int, len(inputArray)) // @step:initialize
	copy(result, inputArray)
	currentIndex := 0 // @step:initialize

	for currentIndex < len(result) {
		currentValue := result[currentIndex]  // @step:compare
		correctIndex := currentValue - 1      // @step:compare

		if correctIndex >= 0 &&
			correctIndex < len(result) &&
			correctIndex != currentIndex &&
			result[correctIndex] != currentValue {
			// @step:compare
			result[currentIndex], result[correctIndex] = result[correctIndex], result[currentIndex] // @step:swap
		} else {
			currentIndex++ // @step:visit
		}
	}

	return result // @step:complete
}
